### PR TITLE
Fix Multi VU deadlock by isolating browser contexts

### PR DESCRIPTION
**Pull request by inancgumus**
## What?

With this fix, the browser module can now work with an **unlimited** number of VUs and iterations (_limited by the machine power and the browser, of course_). As explained in #1112, this PR separates browser contexts to make each iteration focus solely on its pages rather than those created by other iterations. This brings:

- Efficiency (at least 10X depending on the number of VUs)
- Better debugging (less log noise)
- Possibly fewer deadlocks (none seen so far in our tests)

Tested with the script in #971 and saw no deadlocks.

Note: This PR aims to apply this fix with minimum code changes to avoid disrupting the module's working.

## Why?

Please take a look at #1112 and #971 for more details.

We're adding a separate `onAttachedToTarget` hook to act precisely even before the module's event system is used to carry on the `onAttachedToTarget` event. This way, we can _precisely_ control `Connection`'s attachment decisions at the moment that happens. This also prevents other related concurrency issues between `Browser` and `Connection` event management.

In an ideal world, the session management should be outside of `Connection`, and `Browser` would be managing sessions itself. Currently, this is the cleanest way (IMO) to do this.

## The effects of this PR on other open issues

#971: Deadlock: Multi-VU many iterations

There are no longer deadlocks, even with 1000 VUs. See test run ID: 167467.

#966: Creating new page in browser context sometimes times out

```
Uncaught (in promise) GoError: creating new page in browser context: timed out after 30s
	at github.com/grafana/xk6-browser/browser.mapBrowserContext.func4 (native)
	at file:///tmp/TgCEDu/script.js:24:15(8)
 executor=shared-iterations scenario=ui
```

We no longer see these errors. See test run ID: 167467. However, there were plenty in test run ID 167336 (before this fix was applied).

#970: Uncaught (in promise) when navigating due to time out

```
Uncaught (in promise) navigating frame to "https://test.k6.io/": navigating to "https://test.k6.io/": timed out after 30s executor=shared-iterations scenario=ui
```

This still occurs. But this is about the server(s) under the test unable to keep up with the load. When I manually loaded the page on my browser, it gave me 503/502 many times. Here's the test run ID 167467's log:

```
2024-02-20 17:58:07.832 Failed to load resource: the server responded with a status of 502 (Bad Gateway) browser_source=network line_number=0 stacktrace=<nil> url=https://test.k6.io/ 
2024-02-20 17:58:07.835 Failed to load resource: the server responded with a status of 502 (Bad Gateway) browser_source=network line_number=0 stacktrace=<nil> url=https://test.k6.io/
2024-02-20 17:58:07.841 Failed to load resource: the server responded with a status of 502 (Bad Gateway) browser_source=network line_number=0 stacktrace=<nil> url=https://test.k6.io/static/css/site.css
2024-02-20 17:58:07.853 Failed to load resource: the server responded with a status of 502 (Bad Gateway) browser_source=network line_number=0 stacktrace=<nil> url=https://test.k6.io/static/js/prisms.js
2024-02-20 17:58:07.856 Failed to load resource: the server responded with a status of 502 (Bad Gateway) browser_source=network line_number=0 stacktrace=<nil> url=https://test.k6.io/static/css/site.css
2024-02-20 17:58:39.547 Uncaught (in promise) clicking on "a[href=\"/my_messages.php\"]": timed out after 30s executor=per-vu-iterations scenario=browser
2024-02-20 17:58:39.600 Uncaught (in promise) clicking on "a[href=\"/my_messages.php\"]": timed out after 30s executor=per-vu-iterations scenario=browser
2024-02-20 17:58:40.597 Uncaught (in promise) navigating frame to "https://test.k6.io/": navigating to "https://test.k6.io/": timed out after 30s executor=per-vu-iterations scenario=browser
```

Failing here looks normal to me.

## Checklist

- [x] I have performed a self-review of my code
- [x] I have added tests for my changes.
- [x] I have commented on my code, particularly in hard-to-understand areas

## Related PR(s)/Issue(s)

Updates #1112, #971, and #966 (even maybe #970).

### CODE REVIEW

File: common/browser_context.go [hunk 1]
```diff
@@ -289,11 +289,7 @@ func (b *BrowserContext) NewPage() (*Page, error) {
 
 // Pages returns a list of pages inside this browser context.
 func (b *BrowserContext) Pages() []*Page {
-	pages := make([]*Page, 1)
-	for _, p := range b.browser.getPages() {
-		pages = append(pages, p)
-	}
-	return pages
+	return append([]*Page{}, b.browser.getPages()...)
```
Comment (inancgumus) [inline common/browser_context.go:292]: Should this be in another PR :-(

**Review by ankur22 — COMMENTED** [review]
What an amazing outcome from this PR!

Since it touches some areas that are heavily asynchronous and quite complex I've asked some follow up questions to help me understand some of the changes and how they help improve the runtime stability.

File: common/browser.go [hunk 2]
```diff
@@ -123,6 +123,9 @@ func (b *Browser) connect() error {
 
 	b.conn = conn
 
+	// Start the connection to listen for CDP events.
+	conn.start()
```
[the remark this answers is not in the document]
Comment (inancgumus) [reply common/browser.go:127]: I'll answer your other comments, but I have a partial answer for this one: https://github.com/grafana/xk6-browser/pull/1219#discussion_r1497782514. `start()` controls and makes it explicit when to start `Connection`'s background goroutines.

> Could the hook be passed into the NewConnection as an argument, such as:

Yes, but then all the tests would need to be updated with `nil, ' which I didn't prefer to do so because they don't need this hook to work. I can change this part if you think it's important.

File: common/browser.go [hunk 3]
```diff
@@ -224,8 +224,27 @@ func (b *Browser) initEvents() error {
...
+	// (hence it's the first one we have—the default), or the target is in the
+	// same browser context as the current one.
+	//
+	// This allows to attach targets to the same browser context as the current
+	// one, and to the default browser context.
+	//
+	// We don't want to hold the lock for the entire function to prevent
+	// concurrency issues with the browser context being closed while we're
+	// checking it. So, we don't use defer. This is for a future code that might
+	// extend this function, add a defer, and cause this issue to be missed.
```
[the remark this answers is not in the document]
Comment (inancgumus) [reply common/browser.go:241]: Nope, this is about keeping the lock in the shortest amount possible to prevent future regressions from modifications here. Say this method gets modified later; unintentionally, someone adds (maybe me :P) something that holds this lock longer. Since the interactions between these components are complex to figure out. I wanted to make sure that won't happen. I can put this in a closure here and use `defer`. So, the problem isn't about `defer` per se.

File: common/connection.go [hunk 4]
```diff
@@ -131,6 +131,11 @@ type Connection struct {
 	// Reuse the easyjson structs to avoid allocs per Read/Write.
 	decoder jlexer.Lexer
 	encoder jwriter.Writer
+
+	// onTargetAttachedToTarget is called when a new target is attached to the browser.
+	// Returning false will prevent the session from being created.
+	// If onTargetAttachedToTarget is nil, the session will be created.
+	onTargetAttachedToTarget func(*target.EventAttachedToTarget) bool
```
[the remark this answers is not in the document]
Comment (inancgumus) [reply common/connection.go:146]: It's already a mouthful :( I believe `on` in the name and being a function field already reflects that fact. But, I can make it longer :-]

File: common/browser.go [hunk 5]
```diff
@@ -334,6 +334,15 @@ func (b *Browser) isAttachedPageValid(ev *target.EventAttachedToTarget, browserC
 			ev.SessionID, targetPage.TargetID, targetPage.BrowserContextID, browserCtx == nil, targetPage.Type)
 		return false
 	}
+	// If the target is not in the same browser context as the current one, ignore it.
+	// We're not adding a log to prevent overhead.
```
[the remark this answers is not in the document]
Comment (inancgumus) [reply common/browser.go:341]: I forgot this one.

File: common/browser.go [hunk 6]
```diff
@@ -224,8 +224,27 @@ func (b *Browser) initEvents() error {
...
+	// This allows to attach targets to the same browser context as the current
+	// one, and to the default browser context.
+	//
+	// We don't want to hold the lock for the entire function to prevent
+	// concurrency issues with the browser context being closed while we're
+	// checking it. So, we don't use defer. This is for a future code that might
+	// extend this function, add a defer, and cause this issue to be missed.
+	b.contextMu.RLock()
+	defaultContext := b.context == nil
+	if defaultContext || b.context.id == eva.TargetInfo.BrowserContextID {
```
[the remark this answers is not in the document]
Comment (inancgumus) [reply common/browser.go:244]: Since it's in a short context, I don't think that matters a lot. But I can rename it here.

File: common/connection.go [hunk 7]
```diff
@@ -348,7 +348,16 @@ func (c *Connection) recvLoop() {
...
-			c.closeSession(sid, tid)
+			ok := c.closeSession(sid, tid)
+			if !ok {
+				c.logger.Debugf(
+					"Connection:recvLoop:EventDetachedFromTarget",
+					"sid:%v tid:%v wsURL:%q, session not found",
+					sid, tid, c.wsURL,
+				)
+
+				continue
```
[the remark this answers is not in the document]
Comment (inancgumus) [reply common/connection.go:389]: Yes, I did some local checks before adding this `continue` here and have checked all those conditions. 

> If the emit of an event doesn't occur then that could cause a goroutine to block indefinitely such as

That shouldn't happen because we never let the browser attach unrelated pages. This will come from the browser for a page that didn't get attached. This is similar to the answer below.

> how does this work with the actual fix to prevent iterations from tracking each others browserContexts?

Good question :) The answer is the browser will keep sending these events if we don't attach the pages from other browser contexts. Here, we're ignoring its unnecessary demands to improve performance.

File: common/browser.go [hunk 8]
```diff
@@ -224,8 +224,27 @@ func (b *Browser) initEvents() error {
...
 // event. Returning false will stop the event from being processed by the connection.
-func (b *Browser) connectionOnAttachedToTarget(_ *target.EventAttachedToTarget) bool {
-	return true
+func (b *Browser) connectionOnAttachedToTarget(eva *target.EventAttachedToTarget) bool {
+	// Allow connection to attach to target as the browser context is not set
+	// (hence it's the first one we have—the default), or the target is in the
+	// same browser context as the current one.
+	//
+	// This allows to attach targets to the same browser context as the current
+	// one, and to the default browser context.
```
[the remark this answers is not in the document]
Comment (inancgumus) [reply common/browser.go:236]: OK, I can remove that.

File: common/browser.go [hunk 9]
```diff
@@ -334,6 +334,15 @@ func (b *Browser) isAttachedPageValid(ev *target.EventAttachedToTarget, browserC
 			ev.SessionID, targetPage.TargetID, targetPage.BrowserContextID, browserCtx == nil, targetPage.Type)
 		return false
 	}
+	// If the target is not in the same browser context as the current one, ignore it.
+	// We're not adding a log to prevent overhead.
+	if browserCtx.id != targetPage.BrowserContextID {
```
[the remark this answers is not in the document]
Comment (inancgumus) [reply common/browser.go:342]: This and others. Without others, this will work for to an extent :)

File: common/browser.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (ankur22) [reply common/browser.go:241]: How is this different from other areas of the codebase where we work with locks? Don't we fall under the same trap there too? Just trying to understand :)

File: common/browser.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (ankur22) [reply common/browser.go:244]: I see your point, but I'm on the side of renaming it because it will help remove ambiguity quickly when doing a find all for `defaultContext` in the code base.

[the remark this answers is not in the document]
Comment (ankur22) [reply common/browser.go:244]: Or remove the local `defaultContext` and just add `b.context == nil` directly in the if statement.

File: common/browser.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (ankur22) [reply common/browser.go:127]: > I have a partial answer for this one also here: https://github.com/grafana/xk6-browser/pull/1219#discussion_r1497782514. Using start(), we can explicitly control when to start Connection's background goroutines.

We can discuss that in the other comment.

> Yes, but then all the tests would need to be updated with `nil, ' which I didn't prefer to do so because they don't need this hook to work. I can change this part if you think it's important.

If that's the only reason then my preference is to pass it as an argument in `NewConnection`. Otherwise it feels like a hidden feature.

File: common/connection.go [hunk 10]
```diff
@@ -345,6 +345,7 @@ func (c *Connection) recvLoop() {
 				// if a session should be created for the target.
 				ok := c.onTargetAttachedToTarget(eva)
 				if !ok {
+					c.stopWaitingForDebugger(sid)
```
[the remark this answers is not in the document]
Comment (ankur22) [reply common/connection.go:362]: Could you amend the comment on `stopWaitingForDebugger` to include something about how this helps tackle the issue of not needing to attach pages from unrelated contexts?

File: common/connection.go [hunk 11]
```diff
@@ -165,10 +182,13 @@ func NewConnection(ctx context.Context, wsURL string, logger *log.Logger) (*Conn
 		sessions:         make(map[target.SessionID]*Session),
 	}
 
+	return &c, nil
+}
+
+// start starts the connection's main control loop.
+func (c *Connection) start() {
```
[the remark this answers is not in the document]
Comment (inancgumus) [reply common/connection.go:189]: The `http` package is an entirely different case, though. It's a server and should run handlers in different goroutines. However, almost the entire standard library is synchronous. So, I'm not sure about the "plenty" part.

> What advantage does this bring in this scenario?

It doesn't bring much advantage in this case because our whole design needs to change. I wanted to make the first step in that direction. But I'll change this back until we agree on a better design.

**Comment by inancgumus** [comment]
> What an amazing outcome from this PR! 🎉 🥇

Thank you for the enthusiastic feedback! 🙏 ❤️  As detailed in [the discussion](https://github.com/grafana/xk6-browser/issues/1112), the results of this PR were intentionally planned and are a direct outcome of those design choices. Seeing the team's collective efforts pushing our project forward is great. 😊

File: common/browser.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (inancgumus) [reply common/browser.go:127]: See 1746b073.

File: common/browser.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (inancgumus) [reply common/browser.go:241]: Sure :) The other areas are still prone to lock contention; we might want to fix them. I've moved this to defer instead. See b8add515.

File: common/browser.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (inancgumus) [reply common/browser.go:244]: See b8add515.

File: common/browser.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (inancgumus) [reply common/browser.go]: See e0d7db06

File: common/browser.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (inancgumus) [reply common/browser.go]: See https://github.com/grafana/xk6-browser/commit/1297830844088aaec399972ba840f669b11a7619

File: common/connection.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (inancgumus) [reply common/connection.go]: See https://github.com/grafana/xk6-browser/commit/e0d7db06098976fa646f39ef316fcbe081b4aeaf

[the remark this answers is not in the document]
Comment (ankur22) [reply common/connection.go]: Thanks! Happy to discuss this, i'm interested in knowing more from your perspective.

File: common/browser.go [hunk 12]
```diff
@@ -338,7 +334,6 @@ func (b *Browser) isAttachedPageValid(ev *target.EventAttachedToTarget, browserC
 		return false
 	}
 	// If the target is not in the same browser context as the current one, ignore it.
-	// We're not adding a log to prevent overhead.
 	if browserCtx.id != targetPage.BrowserContextID {
 		b.logger.Warnf(
```
Comment (ankur22) [inline common/browser.go:338]: Should this be a debug? It sounds like we're going to get a lot of the logs when running multi VU tests which is only going to add to the noise and confuse users.

Comment (inancgumus) [reply common/browser.go]: Done: c8416b3caf2bf1add6359d78f666c343a1c1124e

**Review by ankur22 — DISMISSED** [review]
LGTM 🚀 🌕

File: common/connection.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (inancgumus) [reply common/connection.go:354]: It's here: [d646e4b](https://github.com/grafana/xk6-browser/pull/1219/commits/d646e4bf16799463ee97246de7b68ebf4cb07d71)

I've also updated another one: [ac1c804](https://github.com/grafana/xk6-browser/pull/1219/commits/ac1c804ecbfc479e10c0340cc0bca0708554220b)

**Review by ankur22 — APPROVED** [review]
